File: stryker4s-poc/src/main/scala/poc/OrderProcessor.scala
package poc

object OrderProcessor {

  final case class Order(quantity: Int, unitPrice: Double)

  def subtotal(order: Order): Double = order.quantity * order.unitPrice

  def discountRate(quantity: Int): Double =
    if (quantity >= 100) 0.20
    else if (quantity >= 50) 0.10
    else if (quantity >= 10) 0.05
    else 0.0

  def total(order: Order): Double = {
    val sub = subtotal(order)
    sub - sub * discountRate(order.quantity)
  }

  def validateOrder(order: Order): Either[String, Order] =
    if (order.quantity <= 0) Left("Quantity must be positive")
    else if (order.unitPrice < 0) Left("Unit price cannot be negative")
    else Right(order)
}

File: strykerjs-node-test-poc/src/orderProcessor.ts
export type Order = { quantity: number; unitPrice: number };

export function subtotal(order: Order): number {
  return order.quantity * order.unitPrice;
}

export function discountRate(quantity: number): number {
  if (quantity >= 100) return 0.2;
  if (quantity >= 50) return 0.1;
  if (quantity >= 10) return 0.05;
  return 0;
}

export function total(order: Order): number {
  const sub = subtotal(order);
  return sub - sub * discountRate(order.quantity);
}

export function validateOrder(order: Order): { ok: true; order: Order } | { ok: false; reason: string } {
  if (order.quantity <= 0) return { ok: false, reason: 'Quantity must be positive' };
  if (order.unitPrice < 0) return { ok: false, reason: 'Unit price cannot be negative' };
  return { ok: true, order };
}
